Source/WebKit/UIProcess/UserMediaProcessManager.cpp

```diff
@@ -188,20 +192,27 @@ bool UserMediaProcessManager::willCreateMediaStream(UserMediaPermissionRequestMa
 void UserMediaProcessManager::revokeSandboxExtensionsIfNeeded(WebProcessProxy& process)
 {
 #if ENABLE(SANDBOX_EXTENSIONS)
+    if (!process.hasAudioCaptureExtension() && !process.hasVideoCaptureExtension())
+        return;
+
     bool hasAudioCapture = false;
     bool hasVideoCapture = false;
     bool hasPendingCapture = false;
 
-    UserMediaPermissionRequestManagerProxy::forEach([&hasAudioCapture, &hasVideoCapture, &hasPendingCapture, &process](auto& managerProxy) {
-        RefPtr proxyPage = managerProxy.page();
-        if (!proxyPage)
-            return;
-        if (&process != &proxyPage->legacyMainFrameProcess())
-            return;
-        hasAudioCapture |= proxyPage->isCapturingAudio();
-        hasVideoCapture |= proxyPage->isCapturingVideo();
-        hasPendingCapture |= managerProxy.hasPendingCapture();
-    });
+    // Check if any page associated with this process has an active or pending capture session.
+    auto pages = process.pages();
+    if (!pages.isEmpty()) {
+        UserMediaPermissionRequestManagerProxy::forEach([pages = WTFMove(pages), &hasAudioCapture, &hasVideoCapture, &hasPendingCapture](auto& managerProxy) mutable {
+            RefPtr proxyPage = managerProxy.page();
+            if (!proxyPage)
+                return;
+            if (!pages.contains(Ref { *proxyPage }))
+                return;
+            hasAudioCapture |= proxyPage->isCapturingAudio();
+            hasVideoCapture |= proxyPage->isCapturingVideo();
```

Comment:
I think this will still be broken because, right now, any process can overwrite the media state set by another process.

In the future, we’ll likely need to move these functions out of `WebPageProxy` so we can track the media state set by each process on the page separately. Then we could also revoke sandbox extensions when an individual process no longer needs them, rather than waiting until all processes associated with a page no longer need them.

Reply:
Hm yeah, I didn't know how `Page::updateIsPlayingMedia` eventually changes state in the media capture flags in `WebPageProxy`, which seems like a pretty big problem.

Reply:
I suspect the minimal change would be to have each `RemotePageProxy` also track its own `mediaState`.

Reply:
But then there's also a bunch of related flags that need fixing, like `activityState` etc.

Reply:
Why does `activityState` need fixing? From when I briefly looked it didn't seem like it was being updated by the WP.